src/go/configgenerator/filter_generator.go

```diff
@@ -0,0 +1,777 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package configgenerator
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/GoogleCloudPlatform/esp-v2/src/go/options"
+	"github.com/GoogleCloudPlatform/esp-v2/src/go/util"
+	"github.com/GoogleCloudPlatform/esp-v2/src/go/util/httppattern"
+	"github.com/golang/glog"
+	"github.com/golang/protobuf/ptypes"
+	anypb "github.com/golang/protobuf/ptypes/any"
+
+	sc "github.com/GoogleCloudPlatform/esp-v2/src/go/configinfo"
+	aupb "github.com/GoogleCloudPlatform/esp-v2/src/go/proto/api/envoy/v9/http/backend_auth"
+	bapb "github.com/GoogleCloudPlatform/esp-v2/src/go/proto/api/envoy/v9/http/backend_auth"
+	commonpb "github.com/GoogleCloudPlatform/esp-v2/src/go/proto/api/envoy/v9/http/common"
+	scpb "github.com/GoogleCloudPlatform/esp-v2/src/go/proto/api/envoy/v9/http/service_control"
+
+	corepb "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	routepb "github.com/envoyproxy/go-control-plane/envoy/config/route/v3"
+	transcoderpb "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/http/grpc_json_transcoder/v3"
+	hcpb "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/http/health_check/v3"
+	jwtpb "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/http/jwt_authn/v3"
+	routerpb "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/http/router/v3"
+	hcmpb "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/network/http_connection_manager/v3"
+	durationpb "github.com/golang/protobuf/ptypes/duration"
+	emptypb "github.com/golang/protobuf/ptypes/empty"
+	wrapperspb "github.com/golang/protobuf/ptypes/wrappers"
+	confpb "google.golang.org/genproto/googleapis/api/serviceconfig"
+	smpb "google.golang.org/genproto/googleapis/api/servicemanagement/v1"
+)
+
+// A wrapper of filter generation logic.
+type FilterGenerator struct {
+	// The filter name.
+	FilterName string
+	// The function to generate filter config and the methods requiring per route configs.
+	FilterGenFunc
+	// The function to generate per route config.
+	// It should be set if the filter needs to set per route config.
+	sc.PerRouteConfigGenFunc
+}
+
+// The function type to generate filter config.
+// Return
+//  - the filter config
+//  - the methods needed to add per route config
+//  - the error
+type FilterGenFunc func(sc *sc.ServiceInfo) (*hcmpb.HttpFilter, []*sc.MethodInfo, error)
+
+// MakeFilterGenerators provide of a slice of FilterGenerator in sequence.
+func MakeFilterGenerators(serviceInfo *sc.ServiceInfo) ([]*FilterGenerator, error) {
```

Comment:
Instead of a slice, can it return a map of filter name to `FilterGenerator`? Then you don't need to store the filter name in the struct, and lookups will be O(1).

Reply:
The lookup overhead is over-concerned since it is less than 10 filters.